File: Solutions/1202. Smallest String With Swaps/1202. Smallest String With Swaps.cpp
class Solution {
public:
    int find(vector<int>& ds ,int i){
        return ds[i]<0?i:ds[i]=find(ds,ds[i]);
    }
    string smallestStringWithSwaps(string s, vector<vector<int>>& pairs) {
        vector<vector<int>> m(s.size());
        vector<int> ds(s.size(),-1);
        for(auto& p:pairs){
            auto i=find(ds,p[0]),j=find(ds,p[1]);
            if(i!=j){
                ds[i]+=ds[j];
                ds[j]=i;
            }
        }
        for(int i=0;i<s.size();++i)
            m[find(ds,i)].push_back(i);
        for(auto& ids:m){
            string ss="";
            for(int id:ids)
                ss+=s[id];
            sort(ss.begin(),ss.end());
            for(int i=0;i<ids.size();++i)
                s[ids[i]]=ss[i];
        }
        return s;
    }
};

File: Solutions/1202. Smallest String With Swaps/1202. Smallest String With Swaps.go
func find(ds *[]int,i int) int {
    if (*ds)[i]<0 {
        return i
    }
    (*ds)[i]=find(ds,(*ds)[i])
    return (*ds)[i]
}

func smallestStringWithSwaps(s string, pairs [][]int) string {
    m:=make([][]int,len(s))
    ds:=make([]int,len(s))
    for i:=0;i<len(s);i+=1{
        ds[i]=-1
    }
    for _,p:=range pairs{
        i,j:=find(&ds,p[0]),find(&ds,p[1])
        if i!=j{
            ds[i]+=ds[j]
            ds[j]=i
        }
    }
    for i:=0;i<len(s);i+=1{
        m[find(&ds,i)]=append(m[find(&ds,i)],i)
    }
    res:=[]byte(s)
    for _,ids:=range m{
        var ss []byte
        for _,id:=range ids{
            ss=append(ss,res[id])
        }
        sort.Slice(res,func(i,j int)bool{return res[i]<res[j]})
        for i:=0;i<len(ids);i+=1{
            res[ids[i]]=ss[i]
        }
    }
    return string(res)
}